File: backend/internal/handlers/revisions.go
package handlers

import (
	"context"
	"errors"
	"net/http"
	"strconv"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/imranp/kerf/backend/internal/middleware"
	"github.com/imranp/kerf/backend/internal/models"
	"github.com/imranp/kerf/backend/internal/tools"
)

// FileRevision is the shape returned by the list/get endpoints. The list
// route returns ContentPreview only; the single-revision route fills Content.
type FileRevision struct {
	ID             string    `json:"id"`
	FileID         string    `json:"file_id"`
	Source         string    `json:"source"`
	UserID         *string   `json:"user_id"`
	UserName       *string   `json:"user_name,omitempty"`
	CreatedAt      time.Time `json:"created_at"`
	Content        *string   `json:"content,omitempty"`
	ContentPreview *string   `json:"content_preview,omitempty"`
}

// RecordRevision inserts a revision row for a file and prunes anything beyond
// `capacity` (oldest evicted). userID is optional — pass nil when the writer
// isn't an authenticated user (e.g. a tool call running outside the request
// auth scope, or a soft-delete with no caller info).
//
// As of Phase 4, storage is base-snapshot + Myers diff with gzip
// compression: every Nth row is a kind='base' row holding the full
// gzipped content, intervening rows store gzipped diffs against their
// parent. The internal heuristics live in `tools.WriteRevision`; this
// HTTP-flavored entrypoint just handles the userID *string → *uuid.UUID
// conversion + non-fatal error semantics callers expect.
//
// This is a best-effort write — we return the error so callers can decide
// whether to abort. In practice, callers log + continue: a missing revision
// row shouldn't fail the user's actual edit.
func RecordRevision(ctx context.Context, pool *pgxpool.Pool, fileID, content, source string, userID *string, capacity int) error {
	if pool == nil || fileID == "" {
		return nil
	}
	var uid *uuid.UUID
	if userID != nil && *userID != "" {
		u, err := uuid.Parse(*userID)
		if err == nil && u != uuid.Nil {
			uid = &u
		}
	}
	_, err := tools.WriteRevision(ctx, pool, fileID, content, source, uid, capacity)
	return err
}

// helper: returns a pointer to userID if non-empty, else nil.
func userIDPtr(uid string) *string {
	if uid == "" {
		return nil
	}
	return &uid
}

// ListRevisions returns the most-recent revisions for a file, with a 200-char
// content preview per row. Newest first. `?limit=` caps the page (default 50,
// max 200).
//
// The preview uses the dedicated content_preview column when present
// (every Phase-4 row writes it on insert) and falls back to a left()
// snapshot of the legacy plaintext column for pre-migration rows. We
// never reconstruct full content here — list-paths must stay O(1) per
// row regardless of base/diff layout.
func (d *Deps) ListRevisions(w http.ResponseWriter, r *http.Request) {
	uid := middleware.UserID(r.Context())
	pid := chi.URLParam(r, "pid")
	fid := chi.URLParam(r, "fid")
	if requireMember(w, r, d.Pool, pid, uid) == "" {
		return
	}
	// Confirm file exists in this project (independent of soft-delete: history
	// is still readable for soft-deleted files so they can be restored).
	var exists bool
	err := d.Pool.QueryRow(r.Context(),
		`select exists(select 1 from files where id = $1 and project_id = $2)`,
		fid, pid).Scan(&exists)
	if err != nil {
		genericServerError(w, err)
		return
	}
	if !exists {
		writeError(w, http.StatusNotFound, "file not found")
		return
	}

	limit := 50
	if v := r.URL.Query().Get("limit"); v != "" {
		if n, err := strconv.Atoi(v); err == nil && n > 0 {
			limit = n
		}
	}
	if limit > 200 {
		limit = 200
	}

	rows, err := d.Pool.Query(r.Context(), `
		select fr.id, fr.file_id, fr.source, fr.user_id, u.name,
		       fr.created_at,
		       coalesce(fr.content_preview, left(fr.content, 200))
		  from file_revisions fr
		  left join users u on u.id = fr.user_id
		 where fr.file_id = $1
		 order by fr.created_at desc
		 limit $2
	`, fid, limit)
	if err != nil {
		genericServerError(w, err)
		return
	}
	defer rows.Close()

	out := []FileRevision{}
	for rows.Next() {
		var (
			rev      FileRevision
			userName *string
			preview  string
		)
		if err := rows.Scan(&rev.ID, &rev.FileID, &rev.Source, &rev.UserID, &userName, &rev.CreatedAt, &preview); err != nil {
			genericServerError(w, err)
			return
		}
		if userName != nil && *userName != "" {
			n := *userName
			rev.UserName = &n
		}
		p := preview
		rev.ContentPreview = &p
		out = append(out, rev)
	}
	writeJSON(w, http.StatusOK, out)
}

// GetRevision returns a single revision with full content, reconstructing
// from base + diffs as needed.
func (d *Deps) GetRevision(w http.ResponseWriter, r *http.Request) {
	uid := middleware.UserID(r.Context())
	pid := chi.URLParam(r, "pid")
	fid := chi.URLParam(r, "fid")
	rid := chi.URLParam(r, "rid")
	if requireMember(w, r, d.Pool, pid, uid) == "" {
		return
	}
	var (
		rev      FileRevision
		userName *string
	)
	// Pull metadata only — content comes from a reconstructor walk.
	err := d.Pool.QueryRow(r.Context(), `
		select fr.id, fr.file_id, fr.source, fr.user_id, u.name, fr.created_at
		  from file_revisions fr
		  left join users u on u.id = fr.user_id
		 inner join files f on f.id = fr.file_id
		 where fr.id = $1 and fr.file_id = $2 and f.project_id = $3
	`, rid, fid, pid).Scan(&rev.ID, &rev.FileID, &rev.Source, &rev.UserID, &userName, &rev.CreatedAt)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			writeError(w, http.StatusNotFound, "revision not found")
			return
		}
		genericServerError(w, err)
		return
	}
	if userName != nil && *userName != "" {
		n := *userName
		rev.UserName = &n
	}
	content, err := tools.ReconstructRevision(r.Context(), d.Pool, rid)
	if err != nil {
		genericServerError(w, err)
		return
	}
	rev.Content = &content
	writeJSON(w, http.StatusOK, rev)
}

// RestoreRevision applies the revision's content to the file (clearing
// deleted_at if set), and inserts a new 'restore' revision row so the restore
// itself is undoable. Returns the updated File.
//
// Reconstructs the target revision via tools.ReconstructRevision; the
// restore row is itself written through the standard RecordRevision
// path, so it lands as a kind='base' (when its parent's run-length is at
// the cap) or as a normal diff against the prior write.
func (d *Deps) RestoreRevision(w http.ResponseWriter, r *http.Request) {
	uid := middleware.UserID(r.Context())
	pid := chi.URLParam(r, "pid")
	fid := chi.URLParam(r, "fid")
	rid := chi.URLParam(r, "rid")
	role := requireMember(w, r, d.Pool, pid, uid)
	if role == "" {
		return
	}
	if role == "viewer" {
		writeError(w, http.StatusForbidden, "viewer cannot restore revisions")
		return
	}

	// Confirm the revision belongs to this file + project before walking
	// the chain.
	var exists bool
	if err := d.Pool.QueryRow(r.Context(), `
		select exists(
		  select 1 from file_revisions fr
		    inner join files f on f.id = fr.file_id
		   where fr.id = $1 and fr.file_id = $2 and f.project_id = $3
		)
	`, rid, fid, pid).Scan(&exists); err != nil {
		genericServerError(w, err)
		return
	}
	if !exists {
		writeError(w, http.StatusNotFound, "revision not found")
		return
	}

	content, err := tools.ReconstructRevision(r.Context(), d.Pool, rid)
	if err != nil {
		genericServerError(w, err)
		return
	}

	// Apply: clear soft-delete, set content, bump updated_at.
	if _, err := d.Pool.Exec(r.Context(), `
		update files set
			content = $1,
			deleted_at = null,
			updated_at = now()
		 where id = $2 and project_id = $3
	`, content, fid, pid); err != nil {
		genericServerError(w, err)
		return
	}

	// Record the restore as a new revision so it's undoable.
	cap := 200
	if d.Cfg != nil && d.Cfg.FileRevisionsMax > 0 {
		cap = d.Cfg.FileRevisionsMax
	}
	if err := RecordRevision(r.Context(), d.Pool, fid, content, "restore", userIDPtr(uid), cap); err != nil {
		// Non-fatal: the file was actually restored. Log via genericServerError
		// would be wrong (it'd 500 a successful restore). Drop silently.
		_ = err
	}

	// Return the updated file.
	d.getFileForResponse(w, r, pid, fid)
}

// getFileForResponse loads + writes a file row using the same shape as GetFile,
// but as an internal helper so handlers can chain it after a mutation.
func (d *Deps) getFileForResponse(w http.ResponseWriter, r *http.Request, pid, fid string) {
	var f models.File
	err := d.Pool.QueryRow(r.Context(), `
		select id, project_id, parent_id, name, kind, content, storage_key, mime_type, size, created_at, updated_at
		  from files where id = $1 and project_id = $2 and deleted_at is null
	`, fid, pid).Scan(&f.ID, &f.ProjectID, &f.ParentID, &f.Name, &f.Kind, &f.Content, &f.StorageKey, &f.MimeType, &f.Size, &f.CreatedAt, &f.UpdatedAt)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			writeError(w, http.StatusNotFound, "file not found")
			return
		}
		genericServerError(w, err)
		return
	}
	d.attachDownloadURL(&f)
	writeJSON(w, http.StatusOK, f)
}

File: packages/kerf-api/src/kerf_api/tools/revisions.py
import json
import uuid
from kerf_chat.tools.registry import ToolSpec, err_payload, ok_payload, register
from kerf_core.utils.context import ProjectCtx
from kerf_core.revisions import (
    reconstruct_revision as _reconstruct_revision,
    write_revision as _write_revision,
)


async def resolve_path(ctx: ProjectCtx, path: str) -> dict:
    clean = path.rstrip("/")
    if not clean.startswith("/"):
        return {"exists": False}
    row = await ctx.pool.fetchrow(
        "SELECT id, parent_id, name, kind FROM files WHERE project_id = $1 AND path = $2 AND deleted_at IS NULL",
        ctx.project_id, clean,
    )
    if not row:
        return {"exists": False}
    return {
        "exists": True,
        "id": row["id"],
        "parent_id": row["parent_id"],
        "name": row["name"],
        "kind": row["kind"],
    }


async def reconstruct_revision(ctx: ProjectCtx, rev_id: uuid.UUID) -> str:
    """Thin wrapper so callers in this module don't need to import kerf_core directly."""
    return await _reconstruct_revision(ctx.pool, rev_id)


async def write_revision(ctx: ProjectCtx, file_id: str, content: str, source: str) -> uuid.UUID:
    """Thin wrapper that threads ctx fields into the canonical implementation."""
    cap = ctx.file_revisions_max if ctx.file_revisions_max > 0 else 200
    user_id = ctx.user_id if ctx.user_id != uuid.UUID(int=0) else None
    return await _write_revision(
        pool=ctx.pool,
        file_id=file_id,
        content=content,
        source=source,
        user_id=user_id,
        cap=cap,
    )


list_revisions_spec = ToolSpec(
    name="list_revisions",
    description="List the most-recent edits to a file as a chronological history (newest first). Returns id, source ('user'|'tool'|'llm'|'restore'), created_at, and a 200-char content_preview per row.",
    input_schema={
        "type": "object",
        "properties": {
            "file_path": {"type": "string"},
            "limit": {"type": "integer"},
        },
        "required": ["file_path"],
    },
)


@register(list_revisions_spec)
async def run_list_revisions(ctx: ProjectCtx, args: bytes) -> str:
    try:
        a = json.loads(args)
    except Exception as e:
        return err_payload(f"invalid args: {e}", "BAD_ARGS")

    file_path = a.get("file_path", "")
    limit = a.get("limit", 50)

    if not file_path:
        return err_payload("file_path is required", "BAD_ARGS")

    if limit <= 0:
        limit = 50
    if limit > 200:
        limit = 200

    rp = await resolve_path(ctx, file_path)
    if not rp.get("exists"):
        fid_row = await ctx.pool.fetchrow(
            "SELECT id FROM files WHERE project_id = $1 AND name = $2 AND deleted_at IS NOT NULL LIMIT 1",
            ctx.project_id, file_path,
        )
        if not fid_row:
            return err_payload(f"file not found: {file_path}", "NOT_FOUND")
        file_id = fid_row["id"]
    else:
        file_id = rp["id"]

    rows = await ctx.pool.fetch(
        """SELECT fr.id, fr.source, fr.user_id, u.name, fr.created_at,
                  COALESCE(fr.content_preview, LEFT(fr.content, 200)) AS content_preview
           FROM file_revisions fr
           LEFT JOIN users u ON u.id = fr.user_id
           WHERE fr.file_id = $1
           ORDER BY fr.created_at DESC
           LIMIT $2""",
        file_id,
        limit,
    )

    revisions = []
    for row in rows:
        user_id_str = str(row["user_id"]) if row["user_id"] else None
        user_name = row["name"] if row["name"] else None
        revisions.append({
            "id": str(row["id"]),
            "source": row["source"],
            "user_id": user_id_str,
            "user_name": user_name,
            "created_at": row["created_at"].isoformat() if row["created_at"] else "",
            "content_preview": row["content_preview"] or "",
        })

    return ok_payload({"revisions": revisions})


restore_revision_spec = ToolSpec(
    name="restore_revision",
    description="Restore a file to one of its previous revisions. Use list_revisions first to find the desired revision id. The restore is itself recorded as a new revision so it can be undone.",
    input_schema={
        "type": "object",
        "properties": {
            "file_path": {"type": "string"},
            "revision_id": {"type": "string"},
        },
        "required": ["file_path", "revision_id"],
    },
)


@register(restore_revision_spec, write=True)
async def run_restore_revision(ctx: ProjectCtx, args: bytes) -> str:
    try:
        a = json.loads(args)
    except Exception as e:
        return err_payload(f"invalid args: {e}", "BAD_ARGS")

    file_path = a.get("file_path", "")
    revision_id_str = a.get("revision_id", "")

    if not file_path or not revision_id_str:
        return err_payload("file_path and revision_id are required", "BAD_ARGS")

    try:
        rev_id = uuid.UUID(revision_id_str)
    except Exception:
        return err_payload("invalid revision_id", "BAD_ARGS")

    rp = await resolve_path(ctx, file_path)
    if not rp.get("exists"):
        fid_row = await ctx.pool.fetchrow(
            "SELECT id FROM files WHERE project_id = $1 AND name = $2 AND deleted_at IS NOT NULL LIMIT 1",
            ctx.project_id, file_path,
        )
        if not fid_row:
            return err_payload(f"file not found: {file_path}", "NOT_FOUND")
        file_id = fid_row["id"]
    else:
        file_id = rp["id"]

    ok_row = await ctx.pool.fetchrow(
        """SELECT EXISTS(
            SELECT 1 FROM file_revisions fr
            INNER JOIN files f ON f.id = fr.file_id
            WHERE fr.id = $1 AND fr.file_id = $2 AND f.project_id = $3
        )""",
        rev_id, file_id, ctx.project_id,
    )
    if not ok_row or not ok_row[0]:
        return err_payload("revision not found", "NOT_FOUND")

    content = await reconstruct_revision(ctx, rev_id)

    await ctx.pool.execute(
        "UPDATE files SET content = $1, deleted_at = NULL, updated_at = now() WHERE id = $2 AND project_id = $3",
        content, file_id, ctx.project_id,
    )

    new_rev = await write_revision(ctx, str(file_id), content, "restore")

    return ok_payload({
        "path": file_path,
        "restored_revision_id": revision_id_str,
        "new_revision_id": str(new_rev),
    })
